docs/openstack/mariadb_copy.md

```diff
@@ -83,31 +93,60 @@ SOURCE_MARIADB_IP=127.17.0.100
 
-  mysql -h $SOURCE_MARIADB_IP -u root "-p$SOURCE_DB_ROOT_PASSWORD" -N -e 'show databases' | while read dbname; do
-      echo "Exporting \$dbname"
+ # Note we do not want to dump the information and performance schema tables so we filter them
+  mysql -h ${SOURCE_MARIADB_IP} -u root "-p${SOURCE_DB_ROOT_PASSWORD}" -N -e 'show databases' | grep -E -v 'schema|mysql' | while read dbname; do
+      echo "Dumping \${dbname}"
       mysqldump -h $SOURCE_MARIADB_IP -uroot "-p$SOURCE_DB_ROOT_PASSWORD" \
           --single-transaction --complete-insert --skip-lock-tables --lock-tables=0 \
-          --databases "\$dbname" \
-          > "\$dbname".sql
+          "\${dbname}" > "\${dbname}".sql
   done
 
   EOF
   ```
 
-* Before restoring the databases from .sql files, we need to ensure that the neutron
-database name is "neutron" and not "ovs_neutron" by running the command:
-
-  ```
-  sed -i '/^CREATE DATABASE/s/ovs_neutron/neutron/g;/^USE/s/ovs_neutron/neutron/g' ovs_neutron.sql
-  ```
-
 * Restore the databases from .sql files into the podified MariaDB:
 
   ```
-  for dbname in cinder glance keystone nova_api nova_cell0 nova ovs_neutron placement; do
-      echo "Importing $dbname"
-      oc run mariadb-client --image $MARIADB_IMAGE -i --rm --restart=Never -- \
-         mysql -h "$PODIFIED_MARIADB_IP" -uroot "-p$PODIFIED_DB_ROOT_PASSWORD" < "$dbname.sql"
+  # db schemas to rename on import
+  declare -A db_name_map
+  db_name_map["nova"]="nova_cell1"
+  db_name_map["ovs_neutron"]="neutron"
+
+  # db servers to import into
+  declare -A db_server_map
+  db_server_map["default"]=${PODIFIED_MARIADB_IP}
+  db_server_map["nova_cell1"]=${PODIFIED_CELL1_MARIADB_IP}
+
+  # db server root password map
+  declare -A db_server_password_map
+  db_server_password_map["default"]=${PODIFIED_DB_ROOT_PASSWORD}
+  db_server_password_map["nova_cell1"]=${PODIFIED_DB_ROOT_PASSWORD}
+
+  all_db_files=$(ls *.sql)
+  for db_file in ${all_db_files}; do
+      db_name=$(echo ${db_file} | awk -F'.' '{ print $1; }')
+      if [[ -v "db_name_map[${db_name}]" ]]; then
+          echo "renaming ${db_name} to ${db_name_map[${db_name}]}"
+          db_name=${db_name_map[${db_name}]}
+      fi
+      db_server=${db_server_map["default"]}
+      if [[ -v "db_server_map[${db_name}]" ]]; then
+          db_server=${db_server_map[${db_name}]}
+      fi
+      db_password=${db_server_password_map["default"]}
+      if [[ -v "db_server_password_map[${db_name}]" ]]; then
+          db_password=${db_server_password_map[${db_name}]}
+      fi
+      echo "creating ${db_name} in ${db_server}"
+      container_name=$(echo "mariadb-client-${db_name}-create" | sed 's/_/-/g')
+      oc run ${container_name} --image ${MARIADB_IMAGE} -i --rm --restart=Never -- \
+          mysql -h "${db_server}" -uroot "-p${db_password}" << EOF
+  CREATE DATABASE IF NOT EXISTS ${db_name} DEFAULT CHARACTER SET ${CHARACTER_SET} DEFAULT COLLATE ${COLLATION};
```

Comment:
If we used the Neutron solution of editing the dump file on "CREATE DATABASE" and "USE" lines, we wouldn't need the explicit DB create and having variables for the char set and collation. Is there a reason not to go for that solution?

Reply:
that is correct but we considered that previously and considered it to be dangerous

sed -i '/^CREATE DATABASE/s/ovs_neutron/neutron/g;/^USE/s/ovs_neutron/neutron/g' ovs_neutron.sql

in principal this will only match lines starting with 
CREATE DATABASE and update ovs_neutron to neutron
and separately match lines starting with USE
and update ovs_neutron to neutron.

the nova db has user data in it and we do not want to assume 
"USE nova" is not in any data the user provides. i.e. cloud init user-data

in nova case we have to rename form nova to nova_cell1
while we likely can craft the regex such that its unlikely to break things we didn't feel comfortable with the neutron approach.
the explicit match of start of line might be sufficient but in general we dint want to take the risk.

i also want to have a large discussion around chartype and coaltation. i think we should be changing the type and normalising it going forward both for greenfild and adoption.

i have not looked at the galarea operator but it need to create
the db server with the default server/database correlation type set
to the same values as are used when adopting.

right now I'm using
```
CARACTER_SET=utf8
COLLATION=utf8_general_ci
```
since that is what happens to be deployed in the standalone cluster

we should likely be using
``` 
CARACTER_SET=utf8mb4
COLLATION=utf8mb4_unicode_ci
```

if your data contains emoji or other multi-bytye unicode
characters you can write the data to a charset of utf8 but it cannot be dumped and restored properly. i.e. it will break db backups.
that is why mysqldump changed its default form utf8 to utf8mb4
https://bugs.mysql.com/bug.php?id=71746

@gibizer what do you think should we recondier the sed approach.
we need to follow up on the green field approach anyway too.
I'm thinking if we are ever going to run an alter table to update
things during adoption is when we should be doing it. i have not tried to implement that as i was going to bring this up in the adoption sync tomorrow

Reply:
I agree that if we need to run alter table then we should do it at import time.


On the sed approach:
```
stack@gibi-devstack-aio:~$ mysqldump --single-transaction --complete-insert --skip-lock-tables --lock-tables=0 --databases nova_api nova_cell1 > dump.sql
stack@gibi-devstack-aio:~$ grep 'INSERT INTO `instance_system_metadata`' dump.sql  | wc
      1     272   18749
stack@gibi-devstack-aio:~$ grep 'INSERT INTO `instance_system_metadata`' dump.sql  | head -c 100 && echo
INSERT INTO `instance_system_metadata` (`created_at`, `updated_at`, `deleted_at`, `id`, `instance_uu
```
So I think the sed that matches on the start of a line in the dump is safe as the insert into tables are a single line starting with INSERT INTO. So if sedding allows us to automatically copying the charset and collection from the source database then it might worth to do the sed.